Leaderboard rows

SheepTester opened the pull request:
<!-- Fill in all spots surrounded by brackets and confirm all check boxes after confirming completion of the tasks -->

# Info

Closes **#41**. (If there is no issue for this pull request yet, please create one or
delete this line if the pull request is for a very minor tweak).

# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

The changes are listed below

## Changes

- Moves leaderboard row to its own component (per #41)

  The props for this component mirror `TopThreeCard`. I personally would've made the component just take a `PublicProfile` object, but taking each property individually would probably be better for creating a placeholder while loading, for example.

- Styles leaderboard row according to design (per #41)

  One deviation I made from the design is left-aligning user names, which I think looks nicer.

  I also had to deviate from the design for the mobile leaderboard so it doesn't look as crammed on mobile. I removed the margins on the sides (so it's flush with the edge of the screen) as well as reducing the spacing. This is what it looked like before the changes:

  <img src="https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/00b75421-eaa6-4868-abce-ca8c472b07ae" alt="Mobile view of design's leaderboard row" height="300">

- Adjusted styles in rest of leaderboard page to match design

- On mobile, the top three leaderboard items wrap like they do on the current leaderboard site. I've added a gap between the top two, which isn't present on the current site:

  Old members site | This PR
  --- | ---
  ![old members leaderboard](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/4e1945c3-bbc7-45c8-a137-d4c747dd4a81) | ![leaderboard in this PR](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/ff5c9228-c32b-4b25-bfe4-40f7e47f7f98)

- Not related to the leaderboard page, but I removed a double scrollbar that appears on the page content. I'm guessing because many of the developers are on Macs, they didn't notice this, but Windows scrollbars are thick 😩
  
  <img src="https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/ee7602c6-a84c-4c63-b70f-09f8d35f8cc7" alt="Double scrollbars" height="300">

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [x] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [x] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [x] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else. *(Vacuously true)*
- [x] My changes produce no new warnings.

# Screenshots

Please include a screenshot of your Cypress testing suite passing successfully.

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/2decbfab-96ca-40c0-99e7-6c140ed088c5)

If you made any visual changes to the website, please include relevant screenshots below.

Desktop | Mobile
--- | ---
![Desktop leaderboard](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/b322780a-9153-41f9-a2ed-9da1da68e7f2) | ![Mobile leaderboard](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/398c5e65-1844-4fc1-919e-d24c65f956e2)

SheepTester committed:
Move leaderboard row to new component

SheepTester committed:
Leaderboard row styling

SheepTester committed:
Mobile leaderboard layout

SheepTester committed:
Improve leaderboard row on mobile

trevorkw7 commented on src/components/leaderboard/LeaderboardRow/index.tsx:
```diff
@@ -0,0 +1,32 @@
+    <div className={styles.row} data-style={position % 2 === 0 ? 'even' : 'odd'}>
+      <span className={styles.position}>{position}</span>
+      <Image
+        src={image}
```
Code Suggestion: Since we're working with the next/image component, it might be worth considering leveraging its auto compression capabilities on the images of these leaderboard rows since every KB counts when we're rendering so many of them on one page. Plus, these images are small so the compression will be less noticeable, yay!  We can do this by using the `quality={}` prop. I tested this on m computer and found that setting the quality to 10/100 dropped the rendered file size by about half from 746kb to 300ish.

trevorkw7 commented on src/components/leaderboard/LeaderboardRow/style.module.scss:
```diff
@@ -0,0 +1,74 @@
+  text-align: right;
+}
+
+@media (max-width: 800px) {
```
Nit: Let's make sure we're using our media breakpoints here defined in `vars.css` so breakpoints can be consistent across the website (so in the future the table will go into mobile mode the same time the navbar will). The vars you'll want to use are: 
`$breakpoint-sm: 425px;
$breakpoint-md: 768px; 
$breakpoint-lg: 1024px;`

trevorkw7 commented on src/components/leaderboard/LeaderboardRow/style.module.scss:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+@media (max-width: 450px) {
```
Same as the previous breakpoint comment

trevorkw7 commented on src/styles/pages/leaderboard.module.scss:
```diff
@@ -1,38 +1,41 @@
+    flex-wrap: wrap;
     gap: 1rem;
     justify-content: space-between;
+
```
Same as the previous breakpoint comment

trevorkw7 commented on src/styles/pages/leaderboard.module.scss:
```diff
@@ -1,38 +1,41 @@
-      &[data-style='odd'] {
-        background-color: var(--theme-background);
-      }
+    @media (max-width: 600px) {
```
Same as the previous breakpoint comment

trevorkw7 commented on src/pages/leaderboard.tsx:
```diff
@@ -23,7 +22,7 @@ const LeaderboardPage = (props: LeaderboardProps) => {
     <div className={styles.container}>
       <div className={styles.header}>
-        <h1>Leaderboard</h1>
+        <h1 className={styles.heading}>Leaderboard</h1>
```
No changes needed for this PR since we need to talk this through at a meeting at some point but it might be a good idea to eventually turn this into a reusable title component for future pages.

trevorkw7 reviewed (changes requested):
Great job with this PR! I really liked how clean your css was and the good structure in the JSX for the LeaderboardRow component. I left a couple comments about using our media breakpoint constants and a little suggestion about using image compression with the next/image component but once we fix those, you might have the record for fastest onboarding to first PR merge in ACM Dev history 😀.

trevorkw7 reviewed (changes requested):
Great job with this PR! I really liked how clean your css was and the good structure in the JSX for the LeaderboardRow component. I left a couple comments about using our media breakpoint constants and a little suggestion about using image compression with the next/image component but once we fix those, you might have the record for fastest onboarding to first PR merge in ACM Dev history 😀.

SheepTester committed:
Use media query breakpoint constants

SheepTester committed:
Reduce leaderboard pfp quality to 10%

by Trevor's suggestion

farisashai commented on src/components/leaderboard/LeaderboardRow/style.module.scss:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+.position {
```
Let's make sure our elements are nested in a matching order to the HTML code.

E.g. HTML Code
```JSX
<div class="row">
    <span class="position" />
    <img class="picture" />
    <div class="nameRank" >
        <span class="name" />
        <span class="rank" />
    </div>
</div>
```

SCSS Code
```SCSS
.row {
    .position {}
    .picture {}
    .nameRank {
        .name {}
        .rank {}
    }
}
```

One of the biggest benefits I've seen from using this properly is that we can easily refactor code into smaller components by dragging that subset of CSS code out at once. It's also nice for code folding in the editor and helping you debug the element or psuedo-state you're selecting on hover with the right extension and it guarantees you don't make a typo anywhere in the structure of the selector so you don't accidentally select the wrong element.

![Screenshot 2023-06-02 at 2 14 17 PM](https://github.com/acmucsd/membership-portal-ui-v2/assets/33165426/a1de468e-dade-4a90-aad5-3117a717a669)

farisashai commented on src/components/leaderboard/LeaderboardRow/style.module.scss:
```diff
@@ -0,0 +1,69 @@
+
+@media (max-width: vars.$breakpoint-md) {
+  .row {
+    gap: 16px;
```
```suggestion
    gap: 1rem;
```
nit: we try to use `rem` values in this project for any values greater than `4px` for better responsiveness on zoom in/out and user font size preferences

farisashai commented on src/components/leaderboard/LeaderboardRow/style.module.scss:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  .position {
+    margin-right: -8px;
```
seeking to learn: what is this necessary for?

SheepTester replied:
Making the row more condensed for mobile so there's more room for the name. The position is left-aligned, so for the low numbers there's an unbalanced amount of whitespace to the right of the number, which I kind of hackily reduce for the mobile view using this negative right margin

An alternative (which would make more sense for numbers tbh) is to right align the position number; I think I'll do that instead now that I think of it

farisashai reviewed (changes requested):
Just left a few more nits - Trevor covered most of them already. Solid progress so far 👍

SheepTester committed:
Reorganize CSS to mirror layout; use rem

SheepTester committed:
Right align position number

Spacing is kind of scuffed due to the varying number lengths

raymosun commented on src/components/leaderboard/LeaderboardRow/style.module.scss:
```diff
@@ -0,0 +1,65 @@
+  align-items: center;
+  background-color: var(--theme-surface-1);
+  display: grid;
+  font-size: 0.9375rem; // 15px per Figma
```
let's consider rounding this to 1 rem for readability

raymosun reviewed (approved):
Looks great 🚀 Just left 1 comment to think about.

SheepTester committed:
Round leaderboard row font size to 1rem

Removing the font-size line because the default font size is 1rem (16px)

SheepTester committed:
Allow leaderboard rows to be taller if text wraps

Now Shravan Hariharan: Factorial Flatbread won't looked crushed anymore, at the cost of occasionally inconsistent leaderboard row heights

SheepTester committed:
Add vertical padding for text-wrapping rows

Consequences of editing GitHub files on the web: all these little commits

SheepTester commented:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/c32a6bb6-18b4-4b36-bdc3-e37092f6e534)
Text only wraps on obscure screen sizes, but when it does I made it look less terrible

raymosun commented:
Looks great! I think this PR is good to merge 🚀

SheepTester commented:
Who is supposed to be the one who merges
